### DISCUSSION TITLE
Getting auth.uid() from within a function

### DISCUSSION THREAD

**Discussion by nruffilo** [supabase/supabase#1114] [Questions] [answered]
I'm attempting to use Store Procedures (well, functions) as a means of making complex queries simple on the UI.  The problem is, I seem to be unable to access auth.uid() from within the functions - which makes them MUCH less secure.  

Here's the Function in question.  The response I get from the front end (which uses .rpc()): "{"hint":null,"details":null,"code":"42501","message":"permission denied for schema auth"}"

Ideally I'd like to use functions as they let me have the front-end simpler, and most important, when doing inserts, it allows me to do multiple inserts in a single query (for example, I have a Quests and UserQuests linking table)

CREATE OR REPLACE FUNCTION GetQuests()
RETURNS TABLE (questId int8, questname varchar, questdescription varchar, questatus int4, reward text, size int4, createddate timestamptz, completeddate timestamptz, expiredate timestamptz)
AS $Body$
  DECLARE userId uuid;
begin
  userId = auth.uid();
  return 
    QUERY SELECT 
            q.QuestId, q.questname, q.questdescription, q.queststatus, q.reward, q.size, q.createddate, q.completeddate, q.expiredate 
          FROM UserQuest uq LEFT JOIN Quest q ON q.questid = uq.questid 
          WHERE uq.user_id = userId;
end;
$Body$
LANGUAGE plpgsql VOLATILE;

**Comment by steve-chavez** [comment, marked as answer]
Hey @nruffilo,

To solve the `permission denied for schema auth` error, try:

```sql
grant usage on schema auth to anon;
grant usage on schema auth to authenticated;
```

That will only allow using the auth schema functions(but not its tables), so it should be safe.

**Reply by nruffilo** [reply]
This worked wonderfully!  Thank you so much, this really opens up tons of new things for me.  I really appreciate your quick and complete response!

**Reply by jasonho-lynx** [reply]
@steve-chavez are there plans to make this access default for new hosted instances? We did it some time back for the docker version:

https://github.com/supabase/supabase/blob/master/docker/dockerfiles/postgres/auth-schema.sql#L93

PR: https://github.com/supabase/supabase/pull/1377
